specification/requirements.adoc

```diff
@@ -1,5 +1,167 @@
+supported version is 1.0.
+
+A CoVE-IO supported device must support the following DOE object types:
+
+|===
+| Vendor ID | Object Type | Description
+
+| 0001 | 00 | DOE Discovery
+| 0001 | 01 | CMA/SPDM
+| 0001 | 02 | Secure CMA/SPDM
+|===
+
+====== Secure Protocol and Data Model (SPDM)
+
+RISC-V CoVE-IO compliant hosts use the Secure Protocol and Data Model (<<SPDM>>)
+protocol to exchange TDISP and IDE Key Management (IDE_KM) messages with
+physical devices, over DOE mailboxes. CoVE-IO also relies on SPDM for
+gathering devices certificates and measurements.
+
+As a consequence, a CoVE-IO supported DSM must support the SPDM specification
+version 1.2 or above. It must also support sending and receiving SPDM Secured
+Messages as defined in the (<<SecuredSPDM>>) specification version 1.1 or above.
+
+TODO: List of required SPDM messages
+
+===== Integrity and Data Exchange (IDE)
+
+The data that flows between a TDI and a TVM must be confidential and both
+integrity and replay attack protected. PCIe IDE provides this protection for all
+Transport Layer Packets (TLPs) moving between those two endpoints. As a
+consequence, CoVE-IO compatible devices must implement and expose this PCIe
+capability.
+
+====== Selective Streams
+
+As all PCIe switches are excluded from a TVM's TCB, CoVE-IO compliant hosts
+TSMs exclusively establish selective IDE streams between the host PCIe Root Port
+and TEE-IO devices.
+
+====== Key Management
+
+To fully support IDE, CoVE-IO compatible devices must implement the IDE Key
+Management (IDE_KM) flows. IDE_KM messages are initiated by the TSM and sent
+over SPDM to the DSM.
+
+=== Host
+
+In order to support TEE-IO devices with the above described capabilities, a
+CoVE implementation must meet software and hardware requirements, as described
+in the next sections.
+
+==== Ownership
+
+In a CoVE-IO context, the non-confidential host software stack is the sole
+platform resources owner. In particular, the physical devices to which TDIs
+belong are owned and managed by the host, typically through the device PF.
+Host software stack components, e.g. the VMM, assign, unassign and expose
+TDIs resp. to, from and to TVMs.
+
+==== Platform Hardware Components
+
+===== PCIe
+
+All CoVE-IO-compliant devices in a CoVE platform must be connected to the CoVE
+platform through a PCIe Root Port (RP) that is part of a PCIe Root Complex (RC).
+
+A CoVE-IO PCIe link can be direct or go via any topology or PCIe switches and
+bridges. Since those are excluded from a TVM's TCB, only selective IDE streams
+must be used between a CoVE-IO device and its corresponding PCIe Root Port.
+
+The TSM establishes one single selective IDE stream for each physical device
+from which TDIs may be attached to TVMs. All TDIs within a CoVE-IO device share
+the same IDE stream.
+
+For a given selective IDE stream, the TSM generates, owns and distributes the
+IDE keys to both the CoVE-IO device and its upstream PCIe Root Port.
```

Comment:
Do we need describe the Root Port capability that only TSM can configure the IDE key, while VMM cannot touch them?

Reply:
I dont think we can enforce that in an implementation agnostic way.
VMM could decide to establish an SPDM session with the ROT and configure IDE keys for a stream ID that then is allocated for CoVE-IO. When TSM comes in and tries to configure IDE keys for the same stream ID, the ROT IDE_KM implementation would reject that request and the TSM would be unable to configure the IDE keys for the given stream ID. Thus it will prevent the TVM from accepting the corresponding TDI.

In the worst case, the TDI can not be assigned to the TVM because the VMM (maliciously or not) already configure IDE keys for the IDE stream that the TSM is supposed to use.

@jyao1 Does that make sense?

Reply:
I think ROT IDE_KM will not reject.
If the device only support on SPDM connection, once TSM will send SPDM GET_VERSION, that behavior will teardown the SPDM connection, and change IDE to insecure. That means everything will restart.

If the device supports multiple SPDM connections, then TSM can setup SPDM without teardown existing SPDM session.

Anyway, I am OK with current text.
And we need add more clear security requirement for HROT to ensure the TDI assignment is free from attack from VMM, because VMM can also setup SPDM at the same time.

Reply:
> I think ROT IDE_KM will not reject. If the device only support one SPDM connection, once TSM will send SPDM GET_VERSION, that behavior will teardown the SPDM connection, and change IDE to insecure. That means everything will restart.
> 

Which is fine I think. TSM can then securely configure the IDE keys from there.

> If the device supports multiple SPDM connections, then TSM can setup SPDM without teardown existing SPDM session.
> 

And that's fine as well. Given that IDE_KM mandates that for a given IDE stream, once a `QUERY` or `KEY_PROG` IDE_KM message has been sent, any `QUERY` or `KEY_PROG` for the same stream coming from a **different** SPDM session will be discarded, we have 2 options:

1. VMM has not configured IDE keys. In that case TSM will be able to do so by first sending `QUERY` and then `KEY_PROG`, etc. After the first `QUERY`, the VMM can no longer set IDE keys for the given stream as long as the TSM keeps the SPDM session opened.
1. VMM has already configure IDE for the IDE stream that TSM is trying to set up. TSM will try to send a `QUERY` message and it will not get a reply and thus will fail to set the IDE keys. TSM tracks that device as being untrusted and must not let the TVM accept it.

Either way, this is fine as well and we don't need anything else from the ROT, besides a valid SPDM and IDE_KM implementation. What do you think?